language-plutus-core/src/Control/Monad/Trans/Inner.hs

```diff
@@ -0,0 +1,64 @@
+-- | Getting monad transormers out of 'Traversable' 'Monad's.
+
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+
+module Control.Monad.Trans.Inner
+    ( InnerT (..)
+    , forBind
+    , yield
+    , mapInnerT
+    ) where
+
+import           Control.Monad
+import           Control.Monad.Except
+
+forBind :: (Monad m, Traversable m, Applicative f) => m a -> (a -> f (m b)) -> f (m b)
+forBind a f = join <$> traverse f a
+
+-- | Turn a 'Traversable' 'Monad' into a monad transformer.
+-- > T Identity   ~ IdentityT
+-- > T Maybe      ~ MaybeT
+-- > T (Either e) ~ ExceptT e
+-- > T ((,) w)    ~ WriterT    -- The tuple is flipped compared to the usual 'WriterT'.
+newtype InnerT f m a = InnerT
+    { unInnerT :: m (f a)
+    }
+
+instance (Functor f, Functor m) => Functor (InnerT f m) where
```

Comment:
This is at least representationally equivalent to `Compose f g`, so you could probably derive a bunch of these via that.